src/App.tsx

```diff
@@ -16,49 +20,14 @@ export const appStore = createStore(appReducer, composedEnhancer);
 
 
 function App() {
+  const componentToExpand = ComponentToTestExpandButton();
```

Comment:
Lepiej nie tworzyć komponentów w taki sposób, a zamiast tego napisać 
`<ComponentToTestExpandButton />`

Wywołanie tego jako funkcji nie jest pożądane z tego powodu, że wtedy react nie rozumie, że to jest component, co powoduje to, że nie da się normalnie użyć hooków wewnątrz tego componentu (takich jak np. `useState`, `useEffect`, etc.)

To jest związane z tym, że trochę dzisiaj sklamałem kiedy mówiłem, że  `React.createElement`, po prostu wywołuje funkcję i przekazuje do niej `propsy`, które my podamy jako argument.
Ta funkcja pod spodem wykonuje trochę więcej rzeczy drobnych rzeczy i co najważniejsze - zwraca klasę `ReactElement`
```js

/**
 * Create and return a new ReactElement of the given type.
 * See https://reactjs.org/docs/react-api.html#createelement
 */
export function createElement(type, config, children) {
.....    
return ReactElement(
    type,
    key,
    ref,
    self,
    source,
    ReactCurrentOwner.current,
    props,
  );
}
````

I ważne w tym wszystkim jest to, że `ReactElement` jest obiektem, z którym można w jakiś sposób połączyć hooki. (nie wiem jak, ponieważ nie zagłębiałem się na razie w to aż tak mocno, ale jakbyś chciała, to możesz zacząć od `348` linii w tym [pliku](https://github1s.com/facebook/react/blob/HEAD/packages/react/src/ReactElement.js) ;) )

W każdym razie, chodzi o to, że React.CreateElement robi trochę więcej niż zwykle wywołanie funkcji.


Podsyłam też artykuł napisany trochę bardziej po ludzku xd

https://medium.com/swlh/never-call-a-react-function-component-6815a92b7856

Reply:
Wow. :O Bardzo dziękuję, super wyjaśnienie. :)